### ISSUE TITLE
v2: Hebbian network - co-access tracking

### ISSUE THREAD

**Issue by 27Bslash6** [27b-io/mcp-memory-service#20]
## Context
Phase 3 of V2 Cognitive Memory Architecture. Depends on schema migration.

## Requirements
Implement Hebbian learning ("fire together, wire together"):
- When multiple memories are returned in a single retrieval, record co-access
- Strengthen edge weight between co-accessed memories: `weight += η` (learning_rate)
- Cap edge weight at 1.0
- Track co_access_count and last_strengthened timestamp

### Parameters (configurable)
- `learning_rate` (η): 0.1
- `min_weight_threshold`: 0.05 (below this, edge gets pruned)

## New API Endpoint
`POST /co-access` - manually record co-access for memory IDs
(Automatic tracking also happens during retrieval)

## Implementation Notes
- Queue co-access events, process in batches (avoid write amplification)
- Store edges in hebbian_edges table (SQLite) or Qdrant payload
- Edge key = sorted tuple of memory IDs (undirected graph)

## Acceptance Criteria
- [ ] Co-access automatically tracked during retrieval
- [ ] Manual co-access endpoint works
- [ ] Edge weights strengthen with repeated co-access
- [ ] Batch processing doesn't block retrieval

**Comment by 27Bslash6** [comment]
## 🧠 Hebbian Learning — Stability, Normalization, and Why Pure Hebb Breaks

This is where the architecture gets genuinely interesting — and where the implementation details matter most.

### The Hebb Rule and Its Problems

Hebb (1949) described the principle qualitatively. The simple mathematical formulation is:
```
Δw_ij = η · x_i · x_j
```

The problem: **this is unstable**. Weights only ever grow (assuming positive activations). In a running system, all edges would eventually saturate at max weight, destroying the information the network encodes. This is known as the "Hebbian catastrophe."

### Oja's Rule — The Fix

Oja (1982) added a normalization term:
```
Δw_ij = η · (x_i · x_j - x_j² · w_ij)
```

The `- x_j² · w_ij` term is subtractive normalization that prevents unbounded growth. It's mathematically equivalent to renormalizing the weight vector after each update.

### BCM Theory — The Better Fix

Bienenstock, Cooper & Munro (1982) proposed a sliding modification threshold θ_M:
- If postsynaptic activity > θ_M → strengthen (LTP)
- If postsynaptic activity < θ_M → weaken (LTD)
- θ_M itself slides based on *average recent activity* (homeostatic)

This is powerful because it means the threshold for "important enough to strengthen" adapts to the base rate of activity in the network. In a busy system with lots of co-accesses, the bar for creating lasting associations rises automatically.

### Current Design Review

The current approach:
```
weight += η (η=0.1), capped at 1.0
```

This is additive Hebbian with a hard cap. It works but has weaknesses:
1. **No decay on unreinforced edges** — edges only grow or stay static (until consolidation prunes them in #22, but that's a blunt instrument)
2. **Additive learning** means the 1st co-access and the 50th co-access contribute equally (+0.1 each)

**Suggestion: Use multiplicative updates for both strengthening AND weakening:**
```python
# Strengthen on co-access (diminishing returns)
weight = weight + η * (1.0 - weight)  # approaches 1.0 asymptotically

# Weaken on non-reinforcement (continuous decay)
weight = weight * (1.0 - decay_rate * Δt)  # exponential decay toward 0
```

This gives you:
- **Diminishing returns** on strengthening (first co-accesses matter most)
- **Continuous competition** between strengthening and decay (only actively reinforced edges survive)
- **No hard cap needed** (asymptotic approach to 1.0)

### 📚 Research
- Hebb, D.O. (1949). *The Organization of Behavior*. Wiley.
- Oja, E. (1982). Simplified neuron model as a principal component analyzer. *J. Math. Biology*, 15, 267-273.
- Bienenstock, E.L., Cooper, L.N. & Munro, P.W. (1982). Theory for the development of neuron selectivity: orientation specificity and binocular interaction in visual cortex. *J. Neuroscience*, 2(1), 32-48.
- Turrigiano, G.G. & Nelson, S.B. (2004). Homeostatic plasticity in the developing nervous system. *Nature Reviews Neuroscience*, 5, 97-107. — On why self-stabilizing learning rules are essential.
- Gerstner, W. & Kistler, W. (2002). *Spiking Neuron Models*. Cambridge University Press. Chapter 10 covers Hebbian learning stability in detail.

### ⚠️ Gotchas
- **The `min_weight_threshold` of 0.05 for pruning is good** — but pruning should be lazy (during consolidation), not eager (during access). Eagerly deleting edges during retrieval adds latency and complexity.
- **Batch processing is non-negotiable**: If a single retrieval returns 10 memories, that's potentially 45 edge updates (10 choose 2). Queue them.
- **Undirected edges (sorted tuple key) is correct**: Co-access is symmetric. But when you add the knowledge graph (#23), those edges *are* directional. Keep the two edge types separate.
- **Scale concern**: With N memories, the potential edge space is O(N²). At 2000+ memories, that's 4M potential edges. The min_weight threshold and consolidation pruning are your scaling safeguards — take them seriously.
- **Cold start**: New memories have no edges. The system needs to bootstrap associations somehow. Consider: when storing a new memory, do a quick vector similarity check against recent memories and pre-seed edges for high-similarity pairs (> 0.85 cosine). This simulates "encoding associations" at storage time.
